integration-tests/environments/react-native/metro.config.js

```diff
@@ -16,40 +16,96 @@
+// it is a direct dependency of our app), we add it to the `blockList` and
+// add a mapping for it. We also add it to the list of watched folders, so
+// that any changes to it are picked up by hot reloading.
+
 const path = require("path");
 const fs = require("fs");
 const exclusionList = require("metro-config/src/defaults/exclusionList");
 
-// Read the localDependencies property of package.json and resolve the relative path values
-const appPackageJson = readJson(__dirname, "package.json");
-const localDependencies = Object.fromEntries(
-  Object.entries(appPackageJson.localDependencies || {}).map(([localName, relativePath]) => [
-    localName,
-    path.resolve(__dirname, relativePath),
-  ]),
-);
-
 function readJson(...pathSegemnts) {
   const filePath = path.join(...pathSegemnts);
   return JSON.parse(fs.readFileSync(filePath, "utf8"));
 }
 
-const watchFolders = Object.values(localDependencies);
+function getLinkedDependencies(packagePath, recursive = true, exclude = new Set()) {
+  const packageJson = readJson(packagePath, "package.json");
+  const nodeModulesPath = path.join(packagePath, "node_modules");
+  const files = fs.readdirSync(nodeModulesPath, { encoding: "utf8", withFileTypes: true });
+  const links = files
+    .filter((f) => f.isSymbolicLink())
+    .map(({ name }) => {
+      const linkPath = fs.realpathSync(path.join(nodeModulesPath, name));
+      const linkPackageJson = readJson(linkPath, "package.json");
+      const peerDependencies = Object.keys(linkPackageJson.peerDependencies || {});
+      return { name, path: linkPath, peerDependencies };
+    });
+  // We're only interested in actual dependencies
+  const dependencyLinks = links.filter(({ name }) => name in packageJson.dependencies && !exclude.has(name));
+  if (recursive) {
+    // Ensure we don't visit this package again
+    exclude.add(packageJson.name);
+    // Traverse all dependencies
+    for (const link of dependencyLinks) {
+      const additionalLinks = getLinkedDependencies(link.path, true, exclude);
+      dependencyLinks.push(...additionalLinks);
+    }
+  }
+  // Return only new links
+  return dependencyLinks;
+}
+
+const linkedDependencies = getLinkedDependencies(__dirname);
+console.log(`Linked dependencies: ${linkedDependencies.map(({ name }) => name).join(", ")}`);
+
+const watchFolders = linkedDependencies.map((pkg) => pkg.path);
 
-// Block any node_module loaded from the local dependency, outside of the app's node_modules
-const blockList = Object.values(localDependencies).map(
-  (localPath) => new RegExp(`^${escape(path.join(localPath, "node_modules"))}\\/.*$`),
-);
+const blockedPaths = [];
+for (const pkg of linkedDependencies) {
+  // Block the links
+  blockedPaths.push(path.join(__dirname, "node_modules", pkg.name));
+  // Block any peer dependencies of the links
+  blockedPaths.push(...pkg.peerDependencies.map((peer) => path.join(pkg.path, "node_modules", peer)));
```

Comment:
`optionalDependencies` here as well

Reply:
I would like to understand your use-case for using `optionalDependencies`. In most cases what I actually want to use is `peerDependencyMeta` to mark individual `peerDependencies` optional: https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta

Reply:
I will attempt to add this to my project and we can go over it when I get it working.  I'm just looking ahead at probable hangups.  I have a devDependency on `react-native` which is an `optionalDependency` for anything using that module.  In that case, I want to resolve any conflicts there.

Reply:
However, it is possible this is just a workaround for my problem.  Maybe there is a smarter way to resolve this.  Lets not hold up this PR with that though